lib/src/carry_save_mutiplier.dart

```diff
@@ -0,0 +1,125 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// carry_save_multiplier.dart
+// Implementation of pipeline multiplier module.
+//
+// 2023 May 29
+// Author: Yao Jing Quek <yao.jing.quek@intel.com>
+//
+
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+
+/// A binary multiplier using carry-save technique with pipelining.
+///
+/// Reduces calculation time and complexity by employing the carry save
+/// multiplier, which splits numbers into smaller components and performs
+/// partial multiplications on each component separately, storing the results
+/// in a compact form. The pipeline enhances performance by breaking down the
+/// multiplication process into sequential stages, allowing for concurrent
+/// execution of multiple operations.
+///
+/// The latency of the carry save multiplier is the sum of the two inputs width
+/// `a` and `b`.
+class CarrySaveMultiplier extends Multiplier {
+  /// The list of the sum from every pipeline stages.
+  late final List<Logic> _sum;
+
+  /// The list pf carry from every pipeline stages.
+  late final List<Logic> _carry;
+
+  /// The final product of the multiplier module.
+  @override
+  Logic get product => output('product');
+
+  /// The [latency] of the carry save multiplier.
+  int get latency => super.a.width + 1;
+
+  /// The pipeline for [CarrySaveMultiplier].
+  late final Pipeline _pipeline;
+
+  /// Construct a [CarrySaveMultiplier] that multiply input [a] and input [b].
+  CarrySaveMultiplier(super.a, super.b,
+      {required Logic clk,
+      required Logic reset,
+      super.name = 'carry_save_multiplier'}) {
+    final portA = addInput('a', a, width: a.width);
```

Comment:
Same here, these should be moved into `Multiplier` and then use `a` and `b` inherited from there